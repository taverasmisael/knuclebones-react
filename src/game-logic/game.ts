import type { Optional } from '../utils/optional'
import { Board, BoardCoordinate } from './board'
import { DiceValue } from './dice'
import {
  createNewGameRunning,
  createNewGameNotStarted,
  GameState,
} from './game-status'
import { PlayerBoardPosition, PlayerId } from './player'
import { canPlayerMove, checkForWinner, makeMove } from '../utils/game-helpers'

export interface KnucklebonesGame {
  getGameStatus: () => GameState
  getBoard: () => Optional<Board>
  start: (players: [PlayerId, PlayerId]) => KnucklebonesGame
  moveTopPlayer: (
    coordinate: BoardCoordinate,
    diceValue: DiceValue
  ) => KnucklebonesGame
  moveBottomPlayer: (
    coordinate: BoardCoordinate,
    diceValue: DiceValue
  ) => KnucklebonesGame
}

export function newGame(gs?: GameState): KnucklebonesGame {
  const gameStatus: GameState = gs || createNewGameNotStarted()
  const move = (
    player: PlayerBoardPosition,
    coords: BoardCoordinate,
    diceValue: DiceValue
  ): KnucklebonesGame => {
    if (!canPlayerMove(player, gameStatus)) throw new Error('Not your turn')
    return newGame(
      checkForWinner(makeMove(player, coords, diceValue, gameStatus))
    )
  }

  return {
    getGameStatus: () => gameStatus,
    getBoard: () => gameStatus.board,
    moveTopPlayer: (cords: BoardCoordinate, diceValue: DiceValue) =>
      move(PlayerBoardPosition.TOP, cords, diceValue),
    moveBottomPlayer: (cords: BoardCoordinate, diceValue: DiceValue) =>
      move(PlayerBoardPosition.BOTTOM, cords, diceValue),
    start: (players: [PlayerId, PlayerId]) =>
      newGame(createNewGameRunning(players)),
  }
}
