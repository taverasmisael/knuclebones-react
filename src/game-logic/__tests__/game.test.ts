import { describe, test, expect } from 'vitest'
import { newGame } from '../game'

describe('Game logic', () => {
  test('a new game has an empty board', () => {
    const actual = newGame()
    expect(actual.getBoard()).toMatchSnapshot()
  })

  test('you can start a new game', () => {
    const actual = newGame().start(['player1', 'player2'])
    expect(actual.getBoard()).toMatchSnapshot()
  })
})
