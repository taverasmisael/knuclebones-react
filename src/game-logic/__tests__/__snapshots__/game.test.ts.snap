// Vitest Snapshot v1

exports[`Game logic > a new game has an empty board 1`] = `"None()"`;

exports[`Game logic > bottom player can make a move 1`] = `"Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-1\\"},{\\"value\\":\\"Some(3)\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})"`;

exports[`Game logic > top player can make a move 1`] = `"Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"Some(3)\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})"`;

exports[`Game logic > you can start a new game 1`] = `"Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})"`;
